Use upstream ocaml-base64, drop custom implementation

mseri opened the pull request:
This is fast, tested and more flexible.

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

Commit:
Use upstream base64, drop custom implementation

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

gaborigloi commented:
Are we comfortable with the small behaviour change that this PR brings: in the old `decode` function, we first strip the whitespace from the input, while `B64.decode` does not do that, if there is whitespace in the string, we get a `Not_found` exception? I assume the strings that we would pass to this function cannot contain any whitespace?

mseri commented:
Well spotted. Let me strip them first.

Commit:
Reintroduce strip_whitespace on decode to preserve the old behaviour

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

mseri commented:
The `strip_whitespace` could probably be replaced by `Str.global_replace (Str.regexp "[\r\n\t ]") "" s` (adding str to the dependencies) but I don't know the performance impact.

gaborigloi commented:
Hmm not sure either. Maybe we could try that in a subsequent PR.

gaborigloi commented:
@mseri I've tested the behaviour of the new and old implementations using [qcheck](https://github.com/c-cube/qcheck/), and the behaviour of `encode` seems to be the same. However, the behaviour of decode is different, don't know whether or not that's an issue:
```
# B1.decode "b";;
Exception: (Invalid_argument "index out of bounds").
# B2.decode "b";;
- : bytes = "l" 
```
(Here `B1` is the old behaviour, and `B2` is the new one.)
Also:
```
utop # B1.decode "B7";;
- : bytes = ""                                       
utop # B2.decode "B7";;
- : bytes = "\007"
```
However, the `base64` program doesn't like either of these:
```
> printf b | base64 -d
base64: invalid input
~> printf B7 | base64  -d
base64: invalid input
```

thomassa commented:
Responding to the suggestion of using `Str.global_replace (Str.regexp "[\r\n\t ]") "" s`
Last time I looked, `Str` regexp operations were not thread-safe, so we shouldn't do it that way.

mseri commented:
@thomassa oh, good to know, thanks!

@gaborigloi I don't like that the results are different, although I have just tried additional `Base64` decoders and they also inconsistent. I trust the `unix` tool more in this case, and the `python` implementation fails in a similar way: 
```python
>>> import base64
>>> base64.b64decode("b")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/base64.py", line 78, in b64decode
    raise TypeError(msg)
TypeError: Incorrect padding
>>> base64.b64decode("B7")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/base64.py", line 78, in b64decode
    raise TypeError(msg)
TypeError: Incorrect padding
```

gaborigloi commented:
@mseri  I've compared the old and new implementations for inputs that `base64` considers valid:
```ocaml
module B3 = struct
let encode s =
let (ic, oc) = Unix.open_process "base64" in
output_string oc s; close_out oc; let out = input_line ic in  (assert (Unix.close_process (ic, oc) = Unix.WEXITED 0)); out
let decode s =
let (ic, oc) = Unix.open_process "base64 -d" in
output_string oc s; close_out oc; let out = input_line ic in (assert (Unix.close_process (ic, oc) = Unix.WEXITED 0)); out
end;;
```
```ocaml
 let test_decode_ignore_failure = QCheck.Test.make ~count:10000 ~name:"test b64 change" QCheck.string (fun s -> try B2.decode s = B3.decode s && B2.decode s = B1.decode s with _ -> true);;
```
qcheck found that all the outputs were the same.

Commit:
Improve performance and reliance to standard

According to sections 3.1 and 3.2 of the RFC4648, base64 implementations
should strip the non-alphabetic characters.

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

mseri commented:
Thanks, I think that being unreliable on invalid input is less problematic, and is not changing the behaviour in the sense that we were already wrong.

I believe it's better to use the upstream implementation, and contribute fixes and improvements there (also because there are variation of our custom Base64 implementation across various of our libraries, and I'd like to move them all to one implementation for ease of maintenance and support)

gaborigloi commented:
My test was incorrect, the ocaml code actually disagrees with `base64` in one case at least:
```ocaml
top # let s = "u5M=ZrbU";;
val s : bytes = "u5M=ZrbU"                    
utop # B3.decode s;;
- : (bytes, Unix.process_status) Result.result = Result.Ok "\187\147f\182\212"              
utop # B2.decode s;;
- : bytes = "\187\147\000f\182\212"         
utop # B1.decode s;;
- : bytes = "\187\147\000f\182\212" 
```
For the record, this is the test I used:
@sharady This is the whole file I've used to test this PR:
```ocaml
(* compile with
   ocamlfind ocamlopt -o progprog -linkpkg -package unix -package base64 -package qcheck -package result filename.ml
*)

let read_channel chan =
  let buf = Buffer.create 20 in
  try
    while true do
      input_char chan |> Buffer.add_char buf
    done;
    Buffer.contents buf
  with End_of_file ->
    close_in chan;
    Buffer.contents buf;;

let run_command cmd stdin =
  let ic, oc = Unix.open_process cmd in
  output_string oc stdin;
  close_out oc;
  try
    let output = read_channel ic in
    let status = Unix.close_process (ic, oc) in
    if status = Unix.WEXITED 0 then Result.Ok output else Result.Error status
  with End_of_file ->
    let status = Unix.close_process (ic, oc) in
    Result.Error status;;


module B1 = struct
<contents of base64.ml>
end;;

module B2 = struct
  let sanitize s =
    (* ignore control characters: see RFC4648.1 and RFC4648.3
     * https://tools.ietf.org/html/rfc4648#section-3
     * Note: \t = \009, \n = \012, \r = \015, \s = \032 *)
    let result = Buffer.create (String.length s) in
    for i = 0 to String.length s - 1 do
      if (String.unsafe_get s i >= '\000' && String.unsafe_get s i <= '\032')
      || String.unsafe_get s i = '\127'
      then ()
      else Buffer.add_char result (String.unsafe_get s i)
    done;
    Buffer.contents result

  let encode = B64.encode
  let decode x = B64.decode (sanitize x)
end;;

module B3 = struct
  let encode = run_command "base64"
  let decode = run_command "base64 -d"
end;;

let test_encode_decode =
  QCheck.Test.make ~count:100000 ~name:"test b64 change" QCheck.string
    (fun s -> s |> B2.encode |> B2.decode = s);;

let test_encode_both =
  QCheck.Test.make ~count:100000 ~name:"test b64 change" QCheck.string
    (fun s -> ((B1.encode s) = (B2.encode s)));;

let test_decode_ignore_failure =
  QCheck.Test.make ~count:1000000 ~name:"test b64 change" QCheck.string
    (fun s ->
       QCheck.assume (s <> "");
       match B3.decode s with
       | Result.Ok x -> print_endline "VALID INPUT"; B2.decode s = B1.decode s && B2.decode s = x
       | Result.Error _ -> QCheck.assume(false); true);;

QCheck_runner.run_tests_main
  [ test_encode_decode
  ; test_encode_decode
  ; test_decode_ignore_failure
  ]
```

gaborigloi reviewed (approved):
Looks good, we should definitely move to upstream libraries wherever it's possible.

mseri commented:
According to the RFC, padding should always be at the end of the string.

gaborigloi commented on src/websockets.ml:
```diff
@@ -21,8 +21,23 @@ module Wsprotocol (IO : Iteratees.Monad) = struct
+     * Note: \t = \009, \n = \012, \r = \015, \s = \032 *)
+    let result = Buffer.create (String.length s) in
+    for i = 0 to String.length s - 1 do
+      if String.unsafe_get s i >= '\000' && String.unsafe_get s i <= '\032'
```
I think it would be nice to surround this with a redundant `( ... )` to make the meaning of the code clearer, even though `&&` already has higher precedence than `||`.

mseri replied:
Sure

gaborigloi commented on src/websockets.ml:
```diff
@@ -21,8 +21,23 @@ module Wsprotocol (IO : Iteratees.Monad) = struct
+  let sanitize s =
+    (* ignore control characters: see RFC4648.1 and RFC4648.3
+     * https://tools.ietf.org/html/rfc4648#section-3
+     * Note: \t = \009, \n = \012, \r = \015, \s = \032 *)
```
Hmm, I'm actually confused by what the RFC says:

>  Implementations MUST reject the encoded data if it contains
   characters outside the base alphabet when interpreting base-encoded
   data, unless the specification referring to this document explicitly
   states otherwise.  Such specifications may instead state, as MIME
   does, that characters outside the base encoding alphabet should
   simply be ignored when interpreting data ("be liberal in what you
   accept").

First it says that we MUST reject it, then it says that we MAY ignore it.

thomassa replied:
I take that to mean that when encountering characters outside the base alphabet, the standard base64 behaviour is to reject the data.

But if you're writing a higher-level specification which _uses_ base64, then in that specification you're allowed to state explicitly some other behaviour, overriding the underlying base64 behaviour.

mseri replied:
Yes, that was my same understanding

gaborigloi commented:
I ran the test with qcheck again, and it succeeded. It's important to note though that 99% of the strings that qcheck generates are invalid base64, so it didn't test too many valid inputs.

Commit:
websockets: Make updated code more readable

Signed-off-by: Marcello Seri <marcello.seri@citrix.com>

mseri commented:
I think the test should not be on random strings, where we already know that different implementations have different behaviours, but on the known invariants: (s |> encode |> decode == s) and (encode1 s == encode2 s if you want).

mseri commented:
Note that I am not adding tests to this PR because there is big testing one currently open and in progress.

gaborigloi commented:
@mseri I've run the following tests, and they all passed:
```ocaml
let test_encode_decode =
  QCheck.Test.make ~count:100000 ~name:"test b64 change" QCheck.string
    (fun s -> s |> B2.encode |> B2.decode = s);;

let test_encode_both =
  QCheck.Test.make ~count:100000 ~name:"test b64 change" QCheck.string
    (fun s -> ((B1.encode s) = (B2.encode s)));;

let test_decode_ignore_failure =
  QCheck.Test.make ~count:1000000 ~name:"test b64 change" QCheck.string
    (fun s ->
       QCheck.assume (s <> "");
       match B3.decode s with
       | Result.Ok x -> print_endline "VALID INPUT"; B2.decode s = B1.decode s && B2.decode s = x
       | Result.Error _ -> QCheck.assume(false); true);;

QCheck_runner.run_tests_main
  [ test_encode_decode
  ; test_encode_decode
  ; test_decode_ignore_failure
  ]
```

Btw would it be beneficial to add some of these qcheck tests to the test pull request you mentioned?

gaborigloi reviewed (approved):
This is a good change. We ignore control characters to preserve the old behaviour and compatibility. The tests have verified that the new code is safe.

mseri commented:
Ping @sharady: you can take inspiration from the test above if you need for the other PR

gaborigloi commented:
Yep, these tests might be useful for https://github.com/xapi-project/wsproxy/pull/6.
